render/render.go

```diff
@@ -83,6 +87,30 @@ func mkWriter(path string) (io.Writer, func(), error) {
 
 /* Template functions */
 
+var regexCanonical = regexp.MustCompile(`^(?P<version>v[0-9.]+)`)
+var regexRevision = regexp.MustCompile(`[^-]+-[^-]+-(?P<revision>[a-fA-Z0-9]+)`)
+
+func CanonicalVersion(in string) string {
+	matches := regexCanonical.FindStringSubmatch(module.CanonicalVersion(in))
+	return regexGroup(regexCanonical, "version", matches)
+}
+
+// Revision returns the hash from version strings following this format: <version>-<timestamp>-<hash>
+// If the string does not match this pattern an empty string is returned.
+func Revision(in string) string {
+	matches := regexRevision.FindStringSubmatch(in)
+	return regexGroup(regexRevision, "revision", matches)
+}
+
+func regexGroup(regex *regexp.Regexp, groupName string, matches []string) string {
```

Comment:
I was surprised that getting a named group wasn't something provided by the regexp package already, it looks like it will be in go 1.15 :/

https://github.com/golang/go/issues/32420

Reply:
Nice. I try to not rely on regexp if I can, but good to know it will become easier to use.